crates/driver/src/domain/competition/solution/trade.rs

```diff
@@ -155,7 +155,75 @@ impl Trade {
+                //
+                // See also [`order::Kind::Limit`].
+                //
+                // Similar to market orders, the executed amounts for limit orders are
+                // calculated using the clearing prices.
+                let sell_price = clearing_prices
+                    .0
+                    .get(&sell.token)
+                    .ok_or(Error::ClearingPriceMissing)?
+                    .to_owned();
+                let buy_price = clearing_prices
+                    .0
+                    .get(&buy.token)
+                    .ok_or(Error::ClearingPriceMissing)?
+                    .to_owned();
+                match side {
+                    order::Side::Buy => Execution {
+                        buy: eth::Asset {
+                            amount: executed.into(),
+                            token: buy.token,
+                        },
+                        sell: eth::Asset {
+                            amount: executed
+                                .0
+                                .checked_mul(buy_price)
+                                .ok_or(Error::Overflow)?
+                                .checked_div(sell_price)
+                                .ok_or(Error::Overflow)?
+                                // Because of how "synthetic" limit orders are constructed as
+                                // explained above, we need to simply increase the executed sell
+                                // price by the surplus fee. We know that the user placed an order
+                                // big enough to cover the surplus fee.
+                                .checked_add(surplus_fee.into())
+                                .ok_or(Error::Overflow)?,
+                            token: sell.token,
+                        },
+                    },
+                    order::Side::Sell => Execution {
+                        sell: eth::Asset {
+                            amount: executed.into(),
+                            token: sell.token,
+                        },
+                        buy: eth::Asset {
+                            amount: executed
+                                .0
+                                // Because of how "synthetic" limit orders are constructed as
+                                // explained above, the solver received the sell amount
+                                // reduced by the surplus fee. That's why we're have to reduce the
+                                // executed amount by the surplus fee when calculating the
+                                // executed buy amount.
+                                .checked_sub(surplus_fee.into())
+                                .ok_or(Error::Overflow)?
+                                .checked_mul(sell_price)
+                                .ok_or(Error::Overflow)?
+                                .checked_ceil_div(&buy_price)
```

Comment:
This is a slight change to the current driver code as that one uses `checked_div()` here. Using `checked_div_ceil()` is actually nicer as users would get 1 wei more rather then less (right?) but I wanted to point it out nonetheless in case someone missed it and would like to object.